c/misra/src/rules/DIR-4-6/PlainNumericalTypeUsedOverExplicitTypedef.ql

```diff
@@ -0,0 +1,53 @@
+/**
+ * @id c/misra/plain-numerical-type-used-over-explicit-typedef
+ * @name DIR-4-6: Do not use plain numerical types over typedefs named after their explicit bit layout
+ * @description Using plain numerical types over typedefs with explicit sign and bit counts may lead
+ *              to confusion on how much bits are allocated for a value.
+ * @kind problem
+ * @precision high
+ * @problem.severity error
+ * @tags external/misra/id/dir-4-6
+ *       external/misra/obligation/advisory
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.BuiltInNumericTypes
+
+class BuiltInNumericType extends BuiltInType {
+  BuiltInNumericType() {
+    /* Exclude the plain char because it does not count as a numeric type */
+    this.(CharType).isExplicitlySigned()
+    or
+    this.(CharType).isExplicitlyUnsigned()
+    or
+    this instanceof BuiltInIntegerType
+    or
+    this instanceof FloatType
+    or
+    this instanceof DoubleType
+    or
+    this instanceof LongDoubleType
+  }
+}
+
+predicate forbiddenBuiltinNumericUsedInDecl(Variable var, string message) {
+  var.getType() instanceof BuiltInNumericType and
+  not var instanceof ExcludedVariable and
+  message = "The type " + var.getType() + " is not a fixed-width numeric type."
+}
+
+predicate forbiddenTypedef(TypedefType typedef, string message) {
```

Comment:
To exclude `using x = y`.

```suggestion
predicate forbiddenTypedef(CTypedefType typedef, string message) {
```

Reply:
Since it's a C rule, I don't think we should narrow it down to `CTypedefType`.

Reply:
This is a MISRA C query, and should only ever be run on C++ code. On C code, I would expect this to have no semantic difference. However, I think the benefit of switching to `CTypedefType` is documentary - the intention is clearer.

Reply:
> and should only ever be run on C++ code.

Uh, you mean C code?

Reply:
Yes, sorry! It should never be run on C++ code and only run on C code